extensions/EntityExtension/src/main/kotlin/com/typewritermc/entity/entries/activity/RandomPatrolActivity.kt

```diff
@@ -49,23 +59,65 @@ class RandomPatrolActivity(
 ) : EntityActivity<ActivityContext>, KoinComponent {
     private var network: RoadNetwork? = null
     private var activity: EntityActivity<in ActivityContext> = IdleActivity(startLocation)
+    private val searchMutex = Mutex()
+    private var searchJob: Job? = null
 
     fun refreshActivity(context: ActivityContext, network: RoadNetwork): TickResult {
+        if (searchMutex.isLocked) return TickResult.CONSUMED
+
         val currentPos = currentPosition.toPosition()
-        val nextNode = network.nodes
+        val candidateNodes = network.nodes
             .filter { (it.position.distanceSqrt(currentPos) ?: Double.MAX_VALUE) <= radiusSquared }
-            .randomOrNull()
-            ?: return TickResult.IGNORED
+            .toMutableList()
+
+        if (candidateNodes.isEmpty()) return TickResult.IGNORED
+
+        val job = Dispatchers.UntickedAsync.launch {
+            searchMutex.withLock {
+                if (!isActive) return@withLock
+                var attemptCount = 0
+                while (candidateNodes.isNotEmpty()) {
+                    val nextNode = candidateNodes.removeAt(candidateNodes.indices.random())
+                    attemptCount++
+
+                    val gps = PointToPointGPS(roadNetwork, {
+                        if (currentPosition.toBukkitLocation().block.isSolid) {
+                            currentPosition.withY { it + 1 }
+                        } else {
+                            currentPosition
+                        }.toPosition()
+                    }) { nextNode.position }
+
+                    try {
+                        val pathResult = gps.findPath()
+                        if (pathResult.isSuccess) {
+                            if (!isActive) return@withLock
+                            activity.dispose(context)
+                            activity = NavigationActivity(gps, currentPosition)
+                            activity.initialize(context)
+                            return@withLock
+                        }
+                    } catch (_: Exception) {
+                        // Try next node
+                    }
+                }
+
+                logger.severe("No reachable nodes found after trying $attemptCount attempts from position: $currentPosition")
+                if (!isActive) return@withLock
+                activity.dispose(context)
+
+                val teleportNode =
+                    network.nodes
+                        .filter { (it.position.distanceSqrt(currentPos) ?: Double.MAX_VALUE) <= radiusSquared }
+                        .randomOrNull().logErrorIfNull("No reachable nodes found")
+                activity = IdleActivity(teleportNode?.position?.toProperty() ?: currentPosition).also {
+                    it.initialize(context)
+                }
+            }
+        }
+        searchJob = job
+        job.invokeOnCompletion { if (searchJob === job) searchJob = null }
```

Comment:
This now has a high level of indentation, and I also fail to easily grasph what is happening here. Likely a good sign to split this up into multiple functions.

Also might it be an idea to add a withTimeout here? To avoid reserving the mutex forever?